### PR TITLE
feat: Torch 2.4 + CUDA 12.1 upgrade

### PR DESCRIPTION
**Pull request by bugra-yilmaz**
This PR enables the new _torch-points-kernels_ release compatible with CUDA 12.1 + points to the CUDA 12.1 releases for other Torch libraries. Also, I updated _torch-geometric_ major version to 2.

There is also some refactoring of the toml file, the new release is therefore called `1.4.6+cu121`.

Newly built, published wheel is here - https://pypi.dev.promaton.ai/simple/torch-points3d.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -16,28 +16,19 @@ omegaconf = "^2.1.1"
...
-torch-points-kernels = { url= "https://pypi.dev.promaton.ai/packages/torch_points_kernels-0.7.1+cu117-cp310-cp310-linux_x86_64.whl" }
-torch-cluster = {url = "https://data.pyg.org/whl/torch-2.0.0%2Bcu117/torch_cluster-1.6.3%2Bpt20cu117-cp310-cp310-linux_x86_64.whl"}
-torch-scatter = {url = "https://data.pyg.org/whl/torch-2.0.0%2Bcu117/torch_scatter-2.1.2%2Bpt20cu117-cp310-cp310-linux_x86_64.whl"}
-torch-sparse = {url = "https://data.pyg.org/whl/torch-2.0.0%2Bcu117/torch_sparse-0.6.18%2Bpt20cu117-cp310-cp310-linux_x86_64.whl"}
-torch-geometric = "^1.7.2"
+torch-points-kernels = { version = "==0.7.2+cu121", source = "promaton" }
+torch-cluster = { version = "==1.6.3+pt24cu121", source = "pyg" }
+torch-scatter = { version = "==2.1.2+pt24cu121", source = "pyg" }
+torch-sparse = { version = "==0.6.18+pt24cu121", source = "pyg" }
+torch-geometric = "^2.6.1"
```
Comment (bugra-yilmaz) [inline pyproject.toml:28]: A major upgrade to `torch-geometric`.

File: torch_points3d/core/data_transform/grid_transform.py [hunk 2]
```diff
@@ -118,7 +118,7 @@ def _process(self, data):
         if "batch" not in data:
             cluster = grid_cluster(coords, torch.tensor([1, 1, 1]))
         else:
-            cluster = voxel_grid(coords, data.batch, 1)
+            cluster = voxel_grid(coords, 1, data.batch)
```
Comment (bugra-yilmaz) [inline torch_points3d/core/data_transform/grid_transform.py:121]: And, this was a breaking change related to `torch-geometric`, argument order was different in the new version - causing failures on the IOS side. Fixed with this one line change.

File: pyproject.toml [hunk 3]
```diff
@@ -16,28 +16,19 @@ omegaconf = "^2.1.1"
...
 
 # Additional constraints on no major version creep
-torch = "~2.0.1"
+torch = "~2.4.1"
```
Comment (griegler) [inline pyproject.toml:31]: Is this compatible with torch `2.0.1` currently used e.g., in acd?

File: poetry.lock [hunk 4]
```diff
@@ -1,5 +1,142 @@
 # This file is automatically @generated by Poetry 1.7.0 and should not be changed by hand.
 
+[[package]]
+name = "aiohappyeyeballs"
```
Comment (griegler) [inline poetry.lock:4]: Did not check this file, but where are the new dependencies coming from?

File: pyproject.toml (hunk 3 above)
Comment (bugra-yilmaz) [reply pyproject.toml:31]: Actually, this PR and the corresponding published wheel targets the Torch 2.4.1 upgrade PR on the IOS side - https://github.com/promaton/ios/pull/2953.

So, they are compatible yes.

File: poetry.lock (hunk 4 above)
Comment (bugra-yilmaz) [reply poetry.lock:4]: Obvious upgrades are `torch@2.4.1` + `torch-geometric@2.6.1`, those should imply the new dependencies.

`aiohappyeyeballs` for instance comes as a dependency of `aiohttp` ([see](https://github.com/promaton/torch-points3d/blob/bugra/torch24-cuda121-upgrade/poetry.lock#L115)), which is a dependency of `torch-geometric` ([see](https://github.com/promaton/torch-points3d/blob/bugra/torch24-cuda121-upgrade/poetry.lock#L1325))

File: pyproject.toml (hunk 1 above)
Comment (SeanC-Promaton) [reply pyproject.toml:28]: I think you're fine tbh; organizational shift more than breaking changes, see [here](https://github.com/pyg-team/pytorch_geometric/releases/tag/2.0.0).

```
PyG (PyTorch Geometric) has been moved from my own personal account [rusty1s](https://github.com/rusty1s) to its own organization account [pyg-team](https://github.com/pyg-team) to emphasize the ongoing collaboration between TU Dortmund University, Stanford University and many great external contributors. With this, we are releasing PyG 2.0
```

File: torch_points3d/core/data_transform/grid_transform.py (hunk 2 above)
Comment (SeanC-Promaton) [reply torch_points3d/core/data_transform/grid_transform.py:121]: Nice catch :+1:  That's a minor change in their logs, non-breaking by changing arg order :upside_down_face:

**Review by SeanC-Promaton — APPROVED** [review]
LGTM :+1:

File: pyproject.toml [hunk 5]
```diff
@@ -16,28 +16,19 @@ omegaconf = "^2.1.1"
 # Why do we modify the requirements?
 # This framework is intended for running common deep learning models for point cloud analysis.
 # The original requirements are quite diverse and are not suited for our purposes (i.e. running KPConv in OSS).
-# Hence, we strip out some of the dependencies that are not needed for our use case.
```
Comment (julianschoep) [inline pyproject.toml:19]: Why is this removed? Also the other lines?

File: torch_points3d/core/data_transform/grid_transform.py (hunk 2 above)
Comment (bugra-yilmaz) [reply torch_points3d/core/data_transform/grid_transform.py:121]: Yeah, spent some time into debugging this, luckily it wasn't that difficulty 😅

File: pyproject.toml (hunk 1 above)
Comment (bugra-yilmaz) [reply pyproject.toml:28]: Ah, good to know!

File: pyproject.toml (hunk 5 above)
Comment (bugra-yilmaz) [reply pyproject.toml:19]: There are some deprecated statements there, and I think the kept part is already telling the story:
```
# Why do we modify the requirements?
# This framework is intended for running common deep learning models for point cloud analysis.
# The original requirements are quite diverse and are not suited for our purposes (i.e. running KPConv in OSS).
```

Comment (julianschoep) [reply pyproject.toml:19]: > We set the versions for torch-{cluster, scatter, sparse} fixed here because you need torch installed before poetry can build these. 
> See https://github.com/rusty1s/pytorch_scatter/issues/252. We have to use specific wheels instead.

Are these statements deprecated?

Comment (bugra-yilmaz) [reply pyproject.toml:19]: Yes kind of, the project is moved to PyG source instead (see [this comment](https://github.com/promaton/torch-points3d/pull/8#discussion_r1784235848) from Sean) + the mentioned issue is closed.

Comment (julianschoep) [reply pyproject.toml:19]: Alright cool :+1:

**Review by julianschoep — APPROVED** [review]

**Review by griegler — APPROVED** [review]
